src/routing_table/mod.rs

```diff
@@ -534,6 +534,44 @@ impl<T: Binary + Clone + Copy + Debug + Default + Hash + Xorable> RoutingTable<T
         (result, None)
     }
 
+    /// Adds the given prefix to the routing table, merging or splitting if necessary. Returns the
+    /// entries that have been dropped.
+    pub fn add_prefix(&mut self, prefix: Prefix<T>) -> Vec<T> {
+        let mut result = vec![];
+        if self.our_prefix == prefix || self.sections.contains_key(&prefix) {
+            return result; // We already have this section: Nothing to do!
+        }
+        // While `prefix` extends an existing entry, split that entry.
+        while let Some(&shorter_pfx) =
+            self.sections
+                .keys()
+                .chain(iter::once(&self.our_prefix))
+                .find(|p| p.is_compatible(&prefix) && p.bit_count() < prefix.bit_count()) {
+            let (dropped_nodes, _opt_our_pfx) = self.split(shorter_pfx);
+            result.extend(dropped_nodes);
+        }
+        // Merge if necessary, then add empty sections to satisfy the invariant.
```

Comment:
This part could use some more comments - it's not immediately obvious why we merge in this case and what the `while` loop does (but the algorithm itself is pretty nice ;) ).